File: src/services/cloudinaryService.ts
import { v2 as cloudinary } from "cloudinary";

import { settings } from "../config/settings.js";

cloudinary.config({
  cloud_name: settings.CLOUDINARY_CLOUD_NAME,
  api_key: settings.CLOUDINARY_API_KEY,
  api_secret: settings.CLOUDINARY_API_SECRET,
  secure: true,
});

export const transformProductImage = async (imageUrl: string): Promise<string> => {
  if (!settings.CLOUDINARY_ENABLED) {
    console.log("[Cloudinary] Disabled — returning raw URL");
    return imageUrl;
  }

  let publicId: string | undefined;

  try {
    const uploadResult = await cloudinary.uploader.upload(imageUrl, {
      folder: "itrack/products",
      overwrite: false,
    });
    publicId = uploadResult.public_id;

    const transformedUrl = cloudinary.url(publicId, {
      resource_type: "image",
      transformation: [
        { width: 288, height: 224, crop: "fit" },
        { effect: "sharpen" },
        { effect: "improve", value: "outdoor:50" },
        { quality: "auto:best", fetch_format: "auto" },
      ],
    });

    return transformedUrl;
  } catch {
    // Fall through to fallback path.
  }

  try {
    if (!publicId) {
      const uploadResult = await cloudinary.uploader.upload(imageUrl, {
        folder: "itrack/products",
        overwrite: false,
      });
      publicId = uploadResult.public_id;
    }

    const fallbackUrl = cloudinary.url(publicId, {
      resource_type: "image",
      transformation: [
        { width: 288, height: 224, crop: "fit" },
        { effect: "sharpen" },
        { effect: "improve", value: "outdoor:50" },
        { quality: "auto:best", fetch_format: "auto" },
      ],
    });

    return fallbackUrl;
  } catch (error) {
    console.error("[Cloudinary] Fallback also failed", error);
    return imageUrl;
  }
};

export const uploadScreenshotForLens = async (screenshotB64: string): Promise<string | null> => {
  if (!settings.CLOUDINARY_ENABLED) {
    console.log("[Cloudinary] Disabled - cannot upload screenshot for Lens");
    return null;
  }

  try {
    const dataUri = `data:image/jpeg;base64,${screenshotB64}`;
    const uploadResult = await cloudinary.uploader.upload(dataUri, {
      folder: "itrack/lens-inputs",
      resource_type: "image",
      overwrite: false,
    });

    if (typeof uploadResult.secure_url === "string" && uploadResult.secure_url.length > 0) {
      return uploadResult.secure_url;
    }

    console.warn("[Cloudinary] Screenshot uploaded but secure_url missing for Lens");
    return null;
  } catch (error) {
    console.warn("[Cloudinary] Screenshot upload for Lens failed", error);
    return null;
  }
};

File: services/cloudinary_service.py
"""
Cloudinary Service
------------------
Transforms product images to be visually competitive with Reels.

PRIMARY:   Upload image → generate looping animation
FALLBACK:  Background removal + smart crop only

Controlled by settings.CLOUDINARY_ENABLED.
If False, raw image URL is returned unchanged (useful during dev).
"""

import cloudinary
import cloudinary.uploader
import cloudinary.api
from config.settings import settings

cloudinary.config(
    cloud_name=settings.CLOUDINARY_CLOUD_NAME,
    api_key=settings.CLOUDINARY_API_KEY,
    api_secret=settings.CLOUDINARY_API_SECRET,
    secure=True,
)


async def transform_product_image(image_url: str) -> str:
    """
    Takes a raw product image URL, applies Cloudinary transforms,
    returns the transformed URL.

    If CLOUDINARY_ENABLED is False, returns the original URL unchanged.
    """
    if not settings.CLOUDINARY_ENABLED:
        print("[Cloudinary] Disabled — returning raw URL")
        return image_url

    try:
        # Upload the product image to Cloudinary
        upload_result = cloudinary.uploader.upload(
            image_url,
            folder="itrack/products",
            overwrite=False,
            resource_type="image",
        )
        public_id = upload_result["public_id"]

        # PRIMARY: Try to generate a short looping animation
        # Uses Cloudinary's AI "zoompan" effect to simulate motion from still image
        animated_url = cloudinary.CloudinaryImage(public_id).build_url(
            transformation=[
                {"effect": "zoompan", "zoom": "1.2", "duration": 3},
                {"effect": "loop"},
                {"background_removal": "cloudinary_ai"},
                {"width": 600, "height": 800, "crop": "fill", "gravity": "auto"},
                {"quality": "auto", "fetch_format": "auto"},
            ],
            resource_type="video",  # Animation output
        )
        print(f"[Cloudinary] Animated URL: {animated_url}")
        return animated_url

    except Exception as e:
        print(f"[Cloudinary] Animation failed: {e} — trying background removal fallback")

    try:
        # FALLBACK: Background removal + smart crop
        fallback_url = cloudinary.CloudinaryImage(public_id).build_url(
            transformation=[
                {"effect": "background_removal"},
                {"width": 600, "height": 800, "crop": "fill", "gravity": "auto"},
                {"quality": "auto", "fetch_format": "auto"},
            ],
        )
        print(f"[Cloudinary] Fallback URL: {fallback_url}")
        return fallback_url

    except Exception as e:
        print(f"[Cloudinary] Fallback also failed: {e} — returning raw URL")
        return image_url
